### PR TITLE
kernel (mostly): locally cache values of _current and _current_cpu

### PR DESCRIPTION
**Pull request by jharris-intel**
See e.g. discussion around #32709. Long story short, the compiler isn't particularly good at the best of times at realizing when it is doing redundant calls.

See e.g. https://godbolt.org/z/6jdzhq for a simple example.

Unfortunately, this often has serious knock-on effects (because the compiler doesn't recognize that the two are the same, it'll also e.g. do multiple comparisons, emit code for the case of `_current` being value A then value B later, etc, etc.).

This mainly affects SMP and is a fairly "soft" failure (in the sense that it reduces performance, generally speaking, but doesn't generally affect correctness); don't worry too much about reintroducing duplication later.

This PR gets the low-hanging fruit - that is, cases where `_current` or `_current_cpu` were unconditionally dereferenced more than once within a code path. It does not attempt to tackle e.g. cases where function A dereferences `_current` then calls inline function B that dereferences `_current` again.

As an added bonus, in some cases this results in moving getting `_current` out of a spinlock.

(There _are_ potential cases where this results in worse code, but only slightly so, and I haven't run across any in practice as of yet. Mainly if the compiler does the function call too early and has to spill other values to stack as a result.)

(There is also the possibility here of introducing errors if the resulting caching is overzealous, e.g. if you cache a `_current_cpu` value across a spin unlock/lock.)

### CODE REVIEW

**Review by nandojve — APPROVED** [review]
esWIFI LGTM

File: kernel/smp.c [hunk 1]
```diff
@@ -17,23 +17,26 @@ static atomic_t start_flag;
 unsigned int z_smp_global_lock(void)
 {
 	unsigned int key = arch_irq_lock();
+	struct k_thread *const current = _current;
```
Comment (andyross) [inline kernel/smp.c:20]: This code is the legacy irq_lock() emulation that is only there for compatibility.  Not worth performance work IMHO.

File: kernel/thread.c [hunk 2]
```diff
@@ -810,25 +812,27 @@ void z_init_thread_base(struct _thread_base *thread_base, int priority,
 FUNC_NORETURN void k_thread_user_mode_enter(k_thread_entry_t entry,
 					    void *p1, void *p2, void *p3)
 {
-	_current->base.user_options |= K_USER;
+	struct k_thread *const current = _current;
+
+	current->base.user_options |= K_USER;
```
Comment (andyross) [inline kernel/thread.c:817]: This gets called exactly once per thread, ever.

**Review by andyross — CHANGES_REQUESTED** [review]
Yeah, you're right, that's too many lines for my taste.  This kind of thing is fragile, and the pattern will degrade over time.  And most of this code is never going to be performance-limiting.

Could we shrink this down to just the genuinely hot paths?  Context switching and syscalls seem pretty justified.  A lot of the rest is just hard to see as worth it, IMHO.

And I'd like to spend a little more time working with the optimizer to see if we can find an explanation for why the const attribute on _current[1] is failing to propagate well, and if there's another construction we can use (maybe a expression statement enclosing a static * const that gets returned) that would get us the optimizations without the source code mess.

Changes like this always look fine when they're merged, but three years from now no one is going to remember what this was for and we'll have grown _current's everywhere again.

[1] _current_cpu is always going to need more hand-attention, because we actually want the compiler NEVER to try to optimize it across a lock release.

**Comment by jharris-intel** [comment]
> Yeah, you're right, that's too many lines for my taste. This kind of thing is fragile, and the pattern will degrade over time. And most of this code is never going to be performance-limiting.

Agreed.

> Could we shrink this down to just the genuinely hot paths? Context switching and syscalls seem pretty justified. A lot of the rest is just hard to see as worth it, IMHO.

Honestly, it'd take me more time to figure out where the genuinely hot paths are then it would for me to do another round of "do this rote transformation everywhere"...

> And I'd like to spend a little more time working with the optimizer to see if we can find an explanation for why the const attribute on _current[1] is failing to propagate well, and if there's another construction we can use (maybe a expression statement enclosing a static * const that gets returned) that would get us the optimizations without the source code mess.

I know (approximately) why it's failing to optimize, but I don't know of any good way to work around it.

The primary issue is this: https://godbolt.org/z/KexPoq (similar to but not identical to the previous example).

(Note: identical issue between `const` and `pure` here.)

The main code in this example is this:

```c
    if (bar() || arch_curr_cpu()->swap_ok) {
        baz();
    }
    qux(arch_curr_cpu()->swap_ok);
```
This results in two calls to `arch_curr_cpu`, because the compiler doesn't want to hoist a conditional call into an unconditional call.

Consider for instance if `arch_curr_cpu` would unconditionally hit UB if executed, but `bar` went `puts("welp"); exit(0)`. Behavior would change if you did the obvious optimization. So the compiler won't do that optimization, as it'd be changing behavior.

Now, if I put in a totally useless statement of `asm(""::"r"(arch_curr_cpu()));` at the start, suddenly the compiler goes "oh, this is always called, so I can optimize", and results in a single call to `arch_curr_cpu()`: https://godbolt.org/z/z8bEP6

(Note: asm because otherwise you run into issues here where adding statements just results in the compiler optimizing out the statement and ignoring the implications thereof.)

*****

Or, to put it another way, what we _want_ is some way to tell the compiler that "hey, you can speculate this call". Const says "if you called this function at least once, you can call it only once", but what we want is "hey, you can call this function even if the original codepath may not have called it".

(Which arguably _is_ an attribute of const - but that's not how gcc / clang treat this currently.)

**Comment by jharris-intel** [comment]
>  would unconditionally hit UB if executed

And yes, I am aware that I am playing fast and loose with "UB" here. I know the compiler can assume no UB for the most part.

**Comment by andyross** [comment]
> Now, if I put in a totally useless statement of asm(""::"r"(arch_curr_cpu())); at the start, suddenly the compiler goes "oh, this is always called, so I can optimize

Now... that's actually sounding like a solution to me.   Instead of moving C code around to conform to an obscure set of optimization rules we'll all forget, we can have a ARCH_PERF_PRIME_CURRENT() macro or something to put at the top of any functions of interest.  And when future developers ask "what is this doing here!?" the documentation will tell them that (1) it's a functional noop and safe to remove and (2) it has the effect of making access to _current faster for the rest of the compiler-visible scope (for some platforms and toolchains, yada yada, ymmv, no refunds).

**Comment by jharris-intel** [comment]
> Now... that's actually sounding like a solution to me. Instead of moving C code around to conform to an obscure set of optimization rules we'll all forget, we can have a ARCH_PERF_PRIME_CURRENT() macro or something to put at the top of any functions of interest. And when future developers ask "what is this doing here!?" the documentation will tell them that (1) it's a functional noop and safe to remove and (2) it has the effect of making access to _current faster for the rest of the compiler-visible scope (for some platforms and toolchains, yada yada, ymmv, no refunds).

...huh. That's a good idea. I'll give it a try and see how it goes.

**Comment by jharris-intel** [comment]
Well, that's annoying.

Most of the gains would come from either `_current_cpu` or a consolidation of the two (i.e. if you read `_current` inside an irq lock you don't need the lock & unlock), but because `_current_cpu` can't be `const` most of the places where it _would_ help it can't consolidate (because e.g. it called some external function).

You currently end up with fun things like this:

```asm
0000000000102802 <z_swap_next_thread>:
{
  [...]
  102825:	65 48 8b 14 25 68 00 	mov    %gs:0x68,%rdx
  [...]
  102851:	e8 36 f9 ff ff       	callq  10218c <k_current_get>
  [...]
  102874:	65 48 8b 0c 25 68 00 	mov    %gs:0x68,%rcx
  [...]
  10289b:	65 48 8b 14 25 68 00 	mov    %gs:0x68,%rdx
  [...]
  102901:	65 48 8b 04 25 68 00 	mov    %gs:0x68,%rax
  [...]
  102910:	65 48 8b 04 25 68 00 	mov    %gs:0x68,%rax
```

With 5 direct gets of `_current_cpu`, and an indirect, even with const/pure.

(As an aside, `k_current_get` is a syscall which is why it wasn't inlined there.)

(And because of the memory clobber required for irq lock and unlock, it couldn't even consolidate between a `_current_cpu` and a `_current` even if it _did_ inline.)

My kingdom for typestates... this sort of thing pops up all over the place in embedded programming.

**Comment by jharris-intel** [comment]
I'm going to close this for now. The low-hanging fruit are in #32887, and I'll revisit once that goes through (if indeed it ever does...)

**Comment by jharris-intel** [comment]
I'm going to close this for now. The low-hanging fruit are in #32887, and I'll revisit once that goes through (if indeed it ever does...)

**Pull request closed without merging**